cpp/src/arrow/dataset/scanner.h

```diff
@@ -76,16 +76,55 @@ class ARROW_DS_EXPORT SimpleScanTask : public ScanTask {
   std::vector<std::shared_ptr<RecordBatch>> record_batches_;
 };
 
-/// \brief Main interface for
+/// \brief Scanner is a materialized scan operation with context and options
+/// bound. A scanner is the class that glues ScanTask, DataFragment,
+/// and DataSource. In python pseudo code, it performs the following:
+///
+///  def Scan():
+///    for source in this.sources_:
+///      for fragment in source.GetFragments(this.options_):
+///        for scan_task in fragment.Scan(this.context_):
+///          yield scan_task
 class ARROW_DS_EXPORT Scanner {
  public:
-  /// \brief Return iterator yielding ScanTask instances to enable
-  /// serial or parallel execution of units of scanning work
+  /// \brief The Scan operator returns a stream of ScanTask. The caller is
+  /// responsible to dispatch/schedule said tasks. Tasks should be safe to run
+  /// in a concurrent fashion and outlive the iterator.
   virtual std::unique_ptr<ScanTaskIterator> Scan() = 0;
 
   virtual ~Scanner() = default;
 };
 
+/// \brief SimpleScanner is a trivial Scanner implementation that flattens
+/// chained iterators.
+///
+/// The returned iterator of SimpleScanner::Scan is a serial blocking
+/// iterator. It will block if any of the following methods blocks:
+///  - Iterator::Next
+///  - DataSource::GetFragments
+///  - DataFragment::Scan
+///
+/// Thus, this iterator is not suited for consumption of sources/fragments
```

Comment:
Do you have an idea how a non-trivial Scanner would work in order to hide latencies?

Reply:
- DataSource can cache the list of directories/files, or force an evaluation at creation time, instead of GetFragments.
- DataFragment::Scan is the hardest. For S3 files, you hope that you can "peek" the metadata and make the partitioning in ScanTask from there, e.g. with parquet, you can in theory only read the metadata (via 2 tiny ReadAt) an dispatch a ScanTask per RowGroup. For CSV, you could have companion metadata file that gives offsets of chunks in bytes. The trick is to defer the high latency into the ScanTask.

As for the scanner itself, you can "emulate" this by giving the Scanner internal threads that consume from DataSources and push to a queue, and have ScanTask that hides a future that reads from said queue or yield an empty iterator if everything was consumed. In practice, I will expect that we focus on the pushing the latency in the DataFragments by deferring downloads.

As for your interest, we'll need to refactor the CSV parser to provide an interface where we expose the Chunks (offset of bytes containing full lines?) and bind one ScanTask per Chunk, assuming that type inference is disabled.